### PR TITLE
fix(common): preserve text fragments during same-path navigation

### PR DESCRIPTION
**Pull request by splincode**
## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/main/contributing-docs/commit-message-guidelines.md
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type

What kind of change does this PR introduce?

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [x] angular.dev application / infrastructure changes
- [ ] Other... Please describe:

## What is the current behavior?

Issue Number: #65119

Chrome's "Copy link to highlight" feature creates URLs with text fragments, for example:

```text
/guide/signals#:~:text=signal%20is%20a-,wrapper,-around%20a%20value
```

On Angular Docs, the native browser highlight appears briefly, but then disappears after the docs app finishes loading. The text fragment is removed from the URL because the app performs URL replacement during initial navigation.

## What is the new behavior?

Angular Docs now preserves browser text fragments during the initial same-path URL replacement.

When the current page was loaded from a text-fragment URL, the custom docs Location implementation updates history.state without passing a replacement URL to the History API. This keeps the browser-owned text fragment intact while still preserving Angular Router state updates and URL change notifications.

The fix also guards against stale navigation performance entries after SPA navigation, while still supporting text fragments combined with an element fragment prefix such as:

```text
/guide/signals#some-heading:~:text=signal%20is%20a
```

##### Root Cause

Text-fragment links were affected in two places:

1. The initial same-path `Location.replaceState` could update browser history in a way that loses the text-fragment URL state.
2. Chrome applies native text-fragment highlights before the docs app finishes client setup. `DocViewer` then replaced the SSR/SSG content with `innerHTML`, removing the text nodes that Chrome highlighted.

##### Fix

- Preserve the initial text-fragment URL state while still updating `history.state` and notifying Location URL listeners.
- Share text-fragment URL detection between routing and docs viewer code.
- On initial text-fragment loads, keep existing SSR/SSG docs content when real content is already present.
- Continue running DocViewer client setup for anchors, examples, copy buttons, tabs, breadcrumbs, icons, and TOC.
- Avoid content animation only when existing content is actually preserved.

### CODE REVIEW

**Review by JeanMeche — CHANGES_REQUESTED** [review]
Looks like that still doesn't fix the issue: https://ng-dev-previews-fw--pr-angular-angular-68491-adev-prev-xr8qehze.web.app/overview#:~:text=Maintained%20by%20a%20dedicated%20team%20at%20Google

**Comment by surajy93** [comment]
FYI: Even if the link works, please verify that the highlights are present, as they may disappear after a reload.

**Comment by erkamyaman** [comment]
I tried it on Firefox, the url seems to be fine but highlight and scroll to the highlight have issues

**Comment by JeanMeche** [comment]
Interesting, Firefox seems indeed to work but mark disapears instantly on chrome.

**Comment by splincode** [comment]
Thanks for checking the preview. URL preservation alone was not enough here. The missing piece was `DocViewer`: Chrome applies the native text-fragment highlight before the docs app completes client setup, then `contentContainer.innerHTML = content` replaces the SSR/SSG text nodes and the native highlight disappears. I updated the PR so the initial text-fragment load preserves existing docs DOM when real SSR/SSG content is present, while still running the normal `DocViewer` client setup.

**Comment by splincode** [comment]
Hi, I've reworked the PR based on the previous feedback. The custom ADev Location workaround was removed, and same-path state updates now go through `Location.updateCurrentEntry()` when the Navigation API is available, with replaceState as a fallback. The PR is also squashed into a single commit. Could you please take another look?

**Comment by JeanMeche** [comment]
Looks like it will also need a rebase for the CI to run.

**Comment by splincode** [comment]
@JeanMeche Sorry, I've updated the PR with the CI fixes.

File: adev/src/app/routing/router_providers.ts [hunk 1]
```diff
@@ -134,7 +134,9 @@ const initializeNavigationAdapter = () => {
     }
     if (e instanceof NavigationStart) {
       intercept = true;
-      window.history.replaceState(window.history.state, '', window.location.href);
+      if (currentTextFragmentUrl(window) === undefined) {
```
Comment (atscott) [inline adev/src/app/routing/router_providers.ts]: AGENT: ```suggestion
    if (e instanceof NavigationStart) {
      intercept = true;
      window.history.replaceState(window.history.state, '');
      intercept = false;
    }
```

If we call `window.history.replaceState(window.history.state, '')` (omitting the URL argument), the browser updates history state without modifying the URL in the address bar, keeping any text fragment intact.

This keeps the navigation adapter behavior consistent across all navigations and eliminates the need to import and check `currentTextFragmentUrl` in this file.

File: adev/shared-docs/components/viewers/docs-viewer/docs-viewer.component.ts [hunk 2]
```diff
@@ -97,19 +100,34 @@ export class DocViewer {
     const contentContainer = this.elementRef.nativeElement;
```
Comment (atscott) [inline adev/shared-docs/components/viewers/docs-viewer/docs-viewer.component.ts:98]: AGENT: Question: Is there any reason to restrict preserving the existing DOM only to text fragment loads? On the initial page render of any SSR/SSG page, `contentContainer` already contains the server-rendered HTML. If `DocViewer` skipped `contentContainer.innerHTML = content` on the initial load whenever valid SSR content is already present, it would naturally fix text fragments and avoid redundant DOM recreation across all initial docs page loads.

**Review by atscott — COMMENTED** [review]
AGENT: Thanks for addressing this! Preserving browser-owned text fragments during initial route state updates is a great fix.

A couple of notes for this PR:
1. **Commit Message Scope:** The commit header is `fix(router): ...`, but the framework changes are in `packages/common` (`BrowserPlatformLocation`) and `adev`. Consider updating the scope to `fix(common)` (or splitting the docs changes into a separate commit).
2. **Navigation Adapter:** In `router_providers.ts`, rather than skipping `replaceState` when a text fragment is present, calling `window.history.replaceState(window.history.state, '')` without the URL parameter will preserve the text fragment while keeping behavior consistent and simplifying the code.

File: adev/src/app/routing/router_providers.ts (hunk 1 above)
Comment (splincode) [reply adev/src/app/routing/router_providers.ts]: Updated as suggested. The adapter now always calls replaceState with state and title only, so it preserves the browser-owned URL without a special text-fragment check.

File: adev/shared-docs/components/viewers/docs-viewer/docs-viewer.component.ts (hunk 2 above)
Comment (splincode) [reply adev/shared-docs/components/viewers/docs-viewer/docs-viewer.component.ts:98]: Good point. I generalized this to preserve any non-empty initial SSR/SSG DOM, then use normal replacement for subsequent content updates. This also let me remove the text-fragment URL detector entirely.

**Comment by atscott** [comment]
[Green TGP](https://fusion2.corp.google.com/presubmit/966633469/OCL:966633469:BASE:967203538:1787150517854:c6bb4455/targets)

**Review by atscott — APPROVED** [review]

**Comment by splincode** [comment]
Hi! Just checking in on this PR. It has a green TGP and one approval now.

Is there anything else I should address on my side, or is it just waiting for the remaining reviews?

Thanks!

File: packages/common/src/location/platform_location.ts [hunk 3]
```diff
@@ -162,7 +162,17 @@ export class BrowserPlatformLocation extends PlatformLocation {
   }
 
   override replaceState(state: any, title: string, url: string): void {
-    this._history.replaceState(state, title, url);
+    const targetUrl = new URL(url, this.href).href;
```
Comment (atscott) [inline packages/common/src/location/platform_location.ts]: AGENT: ```suggestion
    const targetUrl = url === '' ? this.href : new URL(url, this._doc.baseURI).href;
```

Per the HTML spec's shared history push/replace state steps, when `url` is non-empty, `history.replaceState` resolves `url` relative to the document's base URL (`this._doc.baseURI`, which accounts for `<base href>`), rather than `this.href` (`location.href`), and leaves the URL unchanged when `url === ''`.

Using `this._doc.baseURI` (with the `url === ''` guard) ensures `targetUrl` matches the browser's resolution when `<base href>` differs from the current URL path/query and a non-root-relative URL is passed (such as `'#/...'\` from `HashLocationStrategy` or a path/query-relative URL).

Comment (splincode) [reply packages/common/src/location/platform_location.ts]: Updated, thanks!

replaceState() now resolves non-empty relative URLs against this._doc.baseURI, with a separate url === '' case. I also added regression tests for both the <base href> resolution and the empty URL behavior.

**Review by atscott — APPROVED** [review]

**Comment by atscott** [comment]
[TGP](https://fusion2.corp.google.com/presubmit/966633469/OCL:966633469:BASE:986930738:1790192262432:369c780d;groups=PossiblyAlreadyFailing/targets) still green with latest changes. should be good to go.

caretaker note: not without risk still. may want to merge and sync on its own

**Review by JeanMeche — APPROVED** [review]
